File: Esercitazione 0/consumatore.c
#include <stdio.h>
#include <fcntl.h>
#include <stdlib.h>
#include <unistd.h>
#include <string.h>
#define MAX_STRING_LENGTH 256

// consumatore.c e' un filtro
int main(int argc, char* argv[]){

	char *file_in, read_char, buf[MAX_STRING_LENGTH];
	int nread, fd;
	
	if(argc==2){

		while(nread = read(0, &read_char, sizeof(char))) /* Fino ad EOF*/{
			if (nread >= 0 ) {
				if(strrchr(argv[1],read_char)==NULL)putchar(read_char);
			}
			else{
				printf("(PID %d) impossibile leggere dal file sorgente", getpid());
				perror("Errore!");
				exit(3);
			}
		}
	}
	else if(argc==3){

		file_in = argv[2];
		fd = open(file_in, O_RDONLY);
		if (fd<0){
			perror("P0: Impossibile aprire il file.");
			exit(2);
		}

		while(nread = read(fd, &read_char, sizeof(char))) /* Fino ad EOF*/{
			if (nread >= 0 ) {
				if(strrchr(argv[1],read_char)==NULL)putchar(read_char);
			}

			else{
				printf("(PID %d) impossibile leggere dal file %s", getpid(), file_in);
				perror("Errore!");
				close(fd);
				exit(3);
			}
		}
		close(fd);

	}
	else{
		perror("Utilizzo: consumatore <filterprefix> <inputFilename> \n oppure \n consumatore <filterprefix> per stdin");
		exit(1);
	}
	
}

File: Esercitazione_0/Consumatore.java
import java.io.FileNotFoundException;
import java.io.FileReader;
import java.io.IOException;

// Consumatore e' un filtro
public class Consumatore {
	public static void main(String[] args) {
		FileReader r = null;
		char ch;
		int x;
	  
		if(args.length==1) {	//codice replicato per evitare controlli multipli nel ciclo
			
			try {
				
				while ((x = System.in.read()) >= 0) { 
					ch = (char) x;
					if(args[0].indexOf(ch)==-1) System.out.print(ch);
				}
				
			} catch(IOException ex){
				
				System.out.println("Errore di input");
				System.exit(2);
			}
			
		}
		
		else if(args.length==2) {
		
			try {
				r = new FileReader(args[1]);
			} catch(FileNotFoundException e){
				
				System.out.println("File non trovato");
				System.exit(1);
			}
			
			try {
				
				while ((x = r.read()) >= 0) { 
					ch = (char) x;
					if(args[0].indexOf(ch)==-1) System.out.print(ch);
				}
				r.close();
			} catch(IOException ex){
				
				System.out.println("Errore di input");
				System.exit(2);
			}
		}
		
		else {
			System.out.println("Utilizzo: consumatore <filterprefix> <inputFilename> \n oppure \n consumatore <filterprefix> per stdin");
			System.exit(0);
		}
	
}}
